src/proto/h2/server.rs

```diff
@@ -267,8 +274,28 @@ where
                         // Record the headers received
                         ping.record_non_data();
 
-                        let req = req.map(|stream| crate::Body::h2(stream, content_length, ping));
-                        let fut = H2Stream::new(service.call(req), respond);
+                        let is_connect = req.method() == Method::CONNECT;
+                        let (mut parts, stream) = req.into_parts();
+                        let (req, connect_parts) = if !is_connect {
+                            (
+                                Request::from_parts(
+                                    parts,
+                                    crate::Body::h2(stream, content_length, ping),
+                                ),
+                                None,
+                            )
+                        } else {
+                            // FIXME(nox): What happens to the request body? Should we check `content_length`?
```

Comment:
iirc in h1 the body is returned as usual, and should be here too: the case is where you get a response that is *not* upgradable (eg you expected 101 and got 4xx) and you decide not to take the upgrade.

Reply:
Ack, I'll change the code accordingly.

Reply:
Actually, Hyper does return an error if you attempt to respond with a successful status code and a body to a CONNECT request:

https://github.com/hyperium/hyper/blob/dbea7716f157896bf7d2d417be7b4e382e7dc34f/src/proto/h1/role.rs#L322-L327

I'll do the same for h2.

Reply:
Wait never mind, that's about the response, not the request body.

Reply:
No, I think this is my fault - as a client you want to be able to read the response body as a body. For a server reading a CONNECT request, any "body" should be present in the upgraded stream rather than the request, even if you choose not to upgrade the request.

Reply:
I made non-zero request and response bodies be an error condition for upgrades, for now. We can always make Hyper support bodies on those requests later, and h1 already reject those bodies in responses on the server-side.